### PR TITLE
Litex soc gen develop

### PR DESCRIPTION
**Pull request by jwfaye**
Add the possibility to have more than one interface for core communication.

### CODE REVIEW

File: litex/tools/litex_soc_gen.py [hunk 1]
```diff
@@ -62,56 +63,66 @@ def build(self, fragment, build_dir, build_name, **kwargs):
 # LiteX SoC Generator ------------------------------------------------------------------------------
 
 class LiteXSoCGenerator(SoCMini):
-    def __init__(self, name="litex_soc", sys_clk_freq=int(50e6), **kwargs):
+    def __init__(self, name="litex_soc", sys_clk_freq=int(50e6), n_mi=3, n_si=0, **kwargs):
+        assert (n_mi + n_si) < 5
```
Comment (enjoy-digital) [inline litex/tools/litex_soc_gen.py:67]: Is this assert arbitrary? If yes, it's probably better to remove it, if not it's better to put a comment.

File: litex/tools/litex_soc_gen.py [hunk 2]
```diff
@@ -62,56 +63,66 @@ def build(self, fragment, build_dir, build_name, **kwargs):
...
+        self.kwargs = kwargs
```
Comment (enjoy-digital) [inline litex/tools/litex_soc_gen.py:68]: Not sure this is required, it seems to be used to get the bus standard, but you can get it with `self.bus.standard`.

File: litex/tools/litex_soc_gen.py [hunk 3]
```diff
@@ -62,56 +63,66 @@ def build(self, fragment, build_dir, build_name, **kwargs):
...
+
         # Platform ---------------------------------------------------------------------------------
-        platform = Platform(device="", io=get_common_ios())
-        platform.name = name
-        platform.add_extension(get_uart_ios())
+        self.platform = Platform(device="", io=get_common_ios())
```
Comment (enjoy-digital) [inline litex/tools/litex_soc_gen.py:71]: Please don't change code that don't require changes.

File: litex/tools/litex_soc_gen.py [hunk 4]
```diff
@@ -62,56 +63,66 @@ def build(self, fragment, build_dir, build_name, **kwargs):
...
+        self.platform.name = name
+        self.platform.add_extension(get_uart_ios())
 
         # CRG --------------------------------------------------------------------------------------
         self.submodules.crg = CRG(
-            clk = platform.request("clk"),
-            rst = platform.request("rst"),
+            clk = self.platform.request("clk"),
```
Comment (enjoy-digital) [inline litex/tools/litex_soc_gen.py:77]: Same as previous.

File: litex/tools/litex_soc_gen.py [hunk 5]
```diff
@@ -62,56 +63,66 @@ def build(self, fragment, build_dir, build_name, **kwargs):
...
+            rst = self.platform.request("rst"),
         )
 
         # SoC --------------------------------------------------------------------------------------
-        if kwargs["uart_name"] == "serial":
-            kwargs["uart_name"] = "uart"
-        SoCCore.__init__(self, platform, clk_freq=sys_clk_freq, ident=f"LiteX standalone SoC - {name}", **kwargs)
+        if self.kwargs["uart_name"] == "serial":
```
Comment (enjoy-digital) [inline litex/tools/litex_soc_gen.py:82]: Same as previous.

File: litex/tools/litex_soc_gen.py [hunk 6]
```diff
@@ -62,56 +63,66 @@ def build(self, fragment, build_dir, build_name, **kwargs):
...
+            self.kwargs["uart_name"] = "uart"
+        SoCCore.__init__(self, self.platform, clk_freq=sys_clk_freq, ident=f"LiteX standalone SoC - {name}", **kwargs)
 
-        # MMAP Slave Interface ---------------------------------------------------------------------
-        s_bus = {
-            "wishbone" : wishbone.Interface(),
-            "axi-lite" : axi.AXILiteInterface(),
+        for n in range(n_mi):
+            origins=[0xa000_0000, 0xb000_0000, 0xc000_0000, 0xd000_0000]
```
Comment (enjoy-digital) [inline litex/tools/litex_soc_gen.py:87]: If we expose the number of interface, we should also probably allow configuration of base addresses. We can wait to address the other points before looking at it.

File: litex/tools/litex_soc_gen.py [hunk 7]
```diff
@@ -122,17 +133,21 @@ def main():
...
     target_group.add_argument("--build",         action="store_true", help="Build SoC.")
     target_group.add_argument("--sys-clk-freq",  default=int(50e6),   help="System clock frequency.")
+    target_group.add_argument("--n-master-inter",  default=int(2),   help="Number of master interfaces")
+    target_group.add_argument("--n-slave-inter",  default=int(0),   help="Number of slave interfaces.")
+
     builder_args(parser)
     soc_core_args(parser)
     args = parser.parse_args()
-
+    print("Argument dict {}".format(args))
```
Comment (enjoy-digital) [inline litex/tools/litex_soc_gen.py:142]: Please remove debug prints.

File: litex/tools/litex_soc_gen.py [hunk 8]
```diff
@@ -122,17 +133,21 @@ def main():
...
     # SoC.
     soc = LiteXSoCGenerator(
         name         = args.name,
         sys_clk_freq = int(float(args.sys_clk_freq)),
+        n_mi = int(args.n_master_inter),
+        n_si = int(args.n_slave_inter),
         **soc_core_argdict(args)
     )
-
```
Comment (enjoy-digital) [inline litex/tools/litex_soc_gen.py:135]: Please avoid touching formatting when not required.

**Review by enjoy-digital — CHANGES_REQUESTED** [review]
Thanks @jwfaye, this will be useful. We'll maybe rename the parameters and also allow configuration of the exposed regions, but before doing this, can you fix the issues related to changes that are not required.

File: litex/tools/litex_soc_gen.py (hunk 3 above)
Comment (jwfaye) [reply litex/tools/litex_soc_gen.py:71]: Hi, I add the "self" to, set it as an attribute as "platform" is needed in the "add_slave_interface" and "slave_master_interface" methods. Sorry for the delay by the way. I was on vacation.

File: litex/tools/litex_soc_gen.py (hunk 1 above)
Comment (jwfaye) [reply litex/tools/litex_soc_gen.py:67]: I set the "assert" as I was thinking about a limitation in the number of interfaces (4 interfaces per standalone SoC.).

File: litex/tools/litex_soc_gen.py (hunk 2 above)
Comment (jwfaye) [reply litex/tools/litex_soc_gen.py:68]: Okay, I will revise the code and submit another version to you.

**Pull request closed without merging**